### PR TITLE
Implement LLR lowering and the rust side of embedding

### PR DESCRIPTION
**Pull request by hunger**
Technically the two changes are independent of each other, but it is so much easier to test them together that I put them up together for review.

### CODE REVIEW

File: internal/core/items/component_container.rs [hunk 1]
```diff
@@ -37,6 +41,77 @@ pub struct ComponentContainer {
...
+        }
+    }
+
+    fn update_from_factory(self: Pin<&Self>, parent_component: &ComponentWeak, index: usize) {
+        let t = self
+            .component_tracker
+            .borrow_mut()
+            .take()
+            .unwrap_or_else(|| Box::pin(PropertyTracker::default()));
+        t.as_ref().evaluate_as_dependency_root(|| self.component_factory());
```
Comment (ogoffart) [inline internal/core/items/component_container.rs:64]:
```rust
 let product = t.as_ref().evaluate_as_dependency_root(|| self.component_factory()).build();
```

Don't need to call twice component_factory.

File: internal/interpreter/api.rs [hunk 2]
```diff
@@ -626,6 +626,22 @@ impl ComponentDefinition {
         })
     }
 
+    /// Instantiate the component using an existing window.
+    #[doc(hidden)]
+    pub fn embed_into_existing_window(
```
Comment (ogoffart) [inline internal/interpreter/api.rs:631]: We need a nice public API for that.
Would be nice to combine with the other function.

File: internal/core/items/component_container.rs [hunk 3]
```diff
@@ -68,45 +173,56 @@ impl Item for ComponentContainer {
...
     }
 
     fn input_event_filter_before_children(
         self: Pin<&Self>,
         _: MouseEvent,
-        _window_adapter: &Rc<dyn WindowAdapter>,
+        window_adapter: &Rc<dyn WindowAdapter>,
         _self_rc: &ItemRc,
     ) -> InputEventFilterResult {
+        self.ensure_updated(window_adapter.window());
```
Comment (ogoffart) [inline internal/core/items/component_container.rs:195]: That's strange, do we really need that before each event?

File: internal/core/items/component_container.rs [hunk 4]
```diff
@@ -40,10 +56,97 @@ pub struct ComponentContainer {
...
 
+impl ComponentContainer {
+    pub fn ensure_updated(self: Pin<&Self>, window: &Window) {
+        let product_change = self
+            .component_tracker
+            .get()
+            .unwrap()
+            .as_ref()
+            .evaluate_if_dirty(|| self.component_factory())
+            .map(|f| {
```
Comment (ogoffart) [inline internal/core/items/component_container.rs:74]: Should they maybe be `and_then`?
Option<Option> is a bit strange.

**Review by ogoffart — APPROVED** [review]

File: internal/interpreter/api.rs (hunk 2 above)
Comment (tronical) [reply internal/interpreter/api.rs:631]: As mentioned before in chat, I'd rather this method is not introduced as it breaks the laziness of the window adapter creation. There's nothing inherent about embedding requiring the adapter. I think, instead, an embedded component should reach out to its parent component when looking for the adapter.

File: internal/core/items/component_container.rs [hunk 5]
```diff
@@ -37,12 +55,116 @@ pub struct ComponentContainer {
...
+        }
+    }
 }
 
 impl Item for ComponentContainer {
-    fn init(self: Pin<&Self>) {}
+    fn init(self: Pin<&Self>, my_component: &ComponentWeak, my_item_index: usize) {
+        self.my_component.set(my_component.clone()).or(Err(())).unwrap();
+
+        let rc = vtable::VWeak::upgrade(my_component).unwrap();
```
Comment (tronical) [inline internal/core/items/component_container.rs:154]: If you change the signature of init to take a &Rc instead of a weak, then you save yourself this unwrap. Going from rc to weak can't fail.

File: internal/core/items/component_container.rs [hunk 6]
```diff
@@ -51,45 +173,56 @@ impl Item for ComponentContainer {
...
     }
 
     fn key_event(
         self: Pin<&Self>,
         _: &KeyEvent,
-        _window_adapter: &Rc<dyn WindowAdapter>,
+        window_adapter: &Rc<dyn WindowAdapter>,
         _self_rc: &ItemRc,
     ) -> KeyEventResult {
+        self.ensure_updated(window_adapter.window());
```
Comment (tronical) [inline internal/core/items/component_container.rs:225]: Why is this needed for this key event function? It bubbles from inside to outside, no?

File: internal/core/items/component_container.rs (hunk 3 above)
Comment (tronical) [reply internal/core/items/component_container.rs:195]: I wonder, too. Shouldn't this be covered by the item visitation?

File: internal/core/items/component_container.rs [hunk 7]
```diff
@@ -37,12 +55,116 @@ pub struct ComponentContainer {
...
-    fn init(self: Pin<&Self>) {}
+    fn init(self: Pin<&Self>, my_component: &ComponentWeak, my_item_index: usize) {
+        self.my_component.set(my_component.clone()).or(Err(())).unwrap();
+
+        let rc = vtable::VWeak::upgrade(my_component).unwrap();
+
+        // Find my embedding item_tree_index:
+        let pin_rc = vtable::VRc::borrow_pin(&rc);
+        let item_tree = pin_rc.as_ref().get_item_tree();
+        let my_node = item_tree.iter().find(|n| matches!(n, ItemTreeNode::Item { item_array_index, .. } if (*item_array_index as usize) == my_item_index)).unwrap();
```
Comment (tronical) [inline internal/core/items/component_container.rs:159]: Perhaps it would be better if the caller of init iterated over the item tree, instead of the items array and then requiring a linear search here for each component container.

File: internal/interpreter/api.rs (hunk 2 above)
Comment (tronical) [reply internal/interpreter/api.rs:631]: I realize that the laziness issue is minor. But I'm more concerned that we're creating the wrong public factory API here that has a Window parameter not because it has any meaning to our users but because it merely simplifies an implementation detail on our side.

Worse: if users don't pass the Window parameter correctly but accidentally call the create function without window it works but gives funny effects (duplicate adapter).

Comment (tronical) [reply internal/interpreter/api.rs:631]: One option is to give the factory an opaque token argument and the component returned must be built with that token. That's how we could smuggle the window through.

I'd much rather prefer the direction you hinted at though: Adding the necessary internal APIs to component vtable to make this work without bothering users. From my changes with the lazy adapters I think that means `window_adapter()` that returns the adapter and creates if necessary, and `maybe_window_adapter()` that returns one of exists and None of not (needed by Drop/destructor).

Comment (ogoffart) [reply internal/interpreter/api.rs:631]: the problem is that the "lazy" window creation might not be lazy enough.  The init call can query the window, or the user querying a property before returning from the factory.
Maybe that means we should decouple the things that need the window from the window. Such as the scale factor, the dark theme and so on. (ie, these property could be accessed through another mean that the Window)
But that might be out of scope for this patch now.

Comment (tronical) [reply internal/interpreter/api.rs:631]: I agree, those properties could be moved out of the adapter. Perhaps stored in the compo, with accessor. Embedded components could use a two-way binding.
